### ISSUE TITLE
Memoryleak by using ITexture::CreateView

### ISSUE THREAD

**Issue by gujeg** [DiligentGraphics/DiligentCore#621]
I need a texture depth stencil array for shadow maps. If I want to use this as a render target, I have to create a TEXTURE_VIEW_DEPTH_STENCIL texture view.
If I do that and release it after use, a memory leak occurs. There are reasons why I have to keep creating the view each render cycle.
It seems that the memory is freed when the program exits, but that is too late.
Please don't tell me to always use the same one. Technically, that should work, is it a problem with the engine, or am I making a mistake ?
Backend is vulkan.
Many thanks in advance for helpful answers.
The following is some example code:

_void RenderShadowMap(ITexture*pTexture,int layer)
{
        const TextureDesc descSource = pTexture->GetDesc();
        ITextureView* pTextureView = nullptr;
        TextureViewDesc desc;
        desc.MostDetailedMip = 0;
        desc.ViewType = TEXTURE_VIEW_DEPTH_STENCIL;
        desc.TextureDim = RESOURCE_DIM_TEX_2D_ARRAY;
        desc.Format = descSource.Format;
        desc.NumMipLevels = 1; //Currently no mipmap support
        desc.Name = "ArrayDepthRT";
        desc.FirstArraySlice = layer;
        desc.NumArraySlices = 1;
        pTexture->CreateView(desc, &pTextureView);
        pContext->SetRenderTargets(0,nullptr, pTextureView, RESOURCE_STATE_TRANSITION_MODE_TRANSITION);
        //Render anything
        pTextureView->Release();    **//<---- Memoryleak occurs !!!!!**
}_

**Comment by TheMostDiligent** [comment]
How do you tell there is a memory leak?

I modified [Tutorial 13](https://github.com/DiligentGraphics/DiligentSamples/blob/master/Tutorials/Tutorial13_ShadowMap/src/Tutorial13_ShadowMap.cpp#L583C28-L583C35) using your code snippet (with a few fixes) and everything works fine:

```cpp
void Tutorial13_ShadowMap::Render()
{
    // Render shadow map
    ITexture*         pTexture     = m_ShadowMapSRV->GetTexture();
    const TextureDesc descSource   = pTexture->GetDesc();
    ITextureView*     pTextureView = nullptr;
    TextureViewDesc   desc;
    desc.MostDetailedMip = 0;
    desc.ViewType        = TEXTURE_VIEW_DEPTH_STENCIL;
    desc.TextureDim      = RESOURCE_DIM_TEX_2D;
    desc.Format          = descSource.Format;
    desc.NumMipLevels    = 1; //Currently no mipmap support
    desc.Name            = "ArrayDepthRT";
    desc.FirstArraySlice = 0;
    desc.NumArraySlices  = 1;
    pTexture->CreateView(desc, &pTextureView);
    m_pImmediateContext->SetRenderTargets(0, nullptr, pTextureView, RESOURCE_STATE_TRANSITION_MODE_TRANSITION);

    m_pImmediateContext->SetRenderTargets(0, nullptr, pTextureView, RESOURCE_STATE_TRANSITION_MODE_TRANSITION);
    m_pImmediateContext->ClearDepthStencil(pTextureView, CLEAR_DEPTH_FLAG, 1.f, 0, RESOURCE_STATE_TRANSITION_MODE_TRANSITION);
    RenderShadowMap();

    //Render anything
    pTextureView->Release(); //<---- No leaks

    // Bind main back buffer
    // ...
```

**Comment by gujeg** [comment]
Sounds strange, that it seems to work for you without any memory leak.
I render around 6000 shadow maps per second (128x128px), so there is a fairly high throughput.
I monitor the application's memory usage in the Windows task manager, or a Linux Memory Visualizer.
There, around 0.3MB of memory is used per second, so after an hour around 1 GB.
As a cross-check, I created the ITextureView as a member variable, i.e. only created it once, but always wrote to the same layer, and there was no memory usage at all. This only occurred when creating, writing and releasing the ITextureView Object.
I have exactly this behavior on different Windows systems and also Linux, Radeons and NVidea cards.

**Comment by TheMostDiligent** [comment]
Can you modify Tutorial 13 as I posted and check if you see the leaks?

**Comment by gujeg** [comment]
I have changed the code accordingly in the tutorial, and the memory leak is very, very clearly visible here too.
The application's memory runs at the same speed as mine, at about 0.3MB per second.
After about 15 minutes it was about 350MB, constantly increasing, with no end in sight.

![image](https://github.com/user-attachments/assets/1a0500bf-fb6a-4085-a186-1a4e356d6874)

![image](https://github.com/user-attachments/assets/fb065736-9126-42bf-8b20-e5a3feb60ba6)

**Comment by gujeg** [comment]
There is no memory leak under DirectX 11 and 12!
Neither under OpenGL, but the code apparently doesn't work because no shadow is visible.

**Comment by gujeg** [comment]
I have found the reason for the memory leak.
It is the map marked here, which continues to grow endlessly.
I think with this information it should be possible for this bug to be fixed in some way for the next version.
Do you agree? And if so, are you integrated into the team sufficiently to be able to arrange this, or do I have to submit the request somewhere else?

![image](https://github.com/user-attachments/assets/262bff01-ec51-4a67-bef9-59441ecd9885)

**Comment by TheMostDiligent** [comment]
Thanks for the investigation, I will take a look at the issue

**Comment by TheMostDiligent** [comment]
@gujeg The issue is now fixed. Notice though that creating many image views each frame is a correct usage of the API but is certainly suboptimal. The views should at least be cached.
Also note that if you are using Vulkan validation layers, there will be memory growth from the layers.

**Comment by gujeg** [comment]
I absolutely agree with you. My views are also cached, it was only temporarily without a cache. But even with this, every now and then I have to recreate 1000s of views and delete old ones because the texture can change in my framework. Anyway, it's great to hear that this bug has been fixed. Great, quick work on the problem, I can only praise it, thank you for that.
Is the fix included in the master branch version of 2.5.6 and automatically included in the next release?

**Comment by TheMostDiligent** [comment]
The fix is in the master branch. You can safely update the submodules now - no need to wait for the next release.
